hwmon: (pmbus/lt3074) add support for lt3074

cencarna opened the pull request:
## PR Description

This adds hardware monitoring and regulator support for Analog Devices LT3074.
The LT3074 is an ultralow noise, ultrafast, 3A, 5.5V dropout linear regulator with
PMBus serial interface which features telemetry for input/output voltage, bias
voltage, output current, and die temperature.

## PR Type
- [ ] Bug fix (a change that fixes an issue)
- [x] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [x] I have conducted a self-review of my own code changes
- [x] I have tested the changes on the relevant hardware
- [x] I have updated the documentation outside this repo accordingly (if there is the case)

nunojsa commented on drivers/hwmon/pmbus/lt3074.c:
```diff
@@ -0,0 +1,141 @@
+	.probe = lt3074_probe,
+	.id_table = lt3074_id,
+};
+
```
nit: no need for extra line in this case

nunojsa commented on drivers/hwmon/pmbus/lt3074.c:
```diff
@@ -0,0 +1,141 @@
+	{ "lt3074", 0 },
+	{}
+};
+
```
same

nunojsa commented on drivers/hwmon/pmbus/lt3074.c:
```diff
@@ -0,0 +1,141 @@
+	ret = i2c_smbus_read_word_data(client, LT3074_MFR_SPECIAL_ID);
+	if (ret != LT3074_SPECIAL_ID_VALUE) {
+		dev_err(dev, "ID 0x%x did not match the device\n", ret);
+		return -ENODEV;
```
`dev_err_probe()`

nunojsa commented on drivers/hwmon/pmbus/lt3074.c:
```diff
@@ -0,0 +1,141 @@
+		break;
+	default:
+		ret = -ENODATA;
+		break;
```
just `return -ENODATA`. Same in the other cases and then no need for `ret`

nunojsa commented on drivers/hwmon/pmbus/lt3074.c:
```diff
@@ -0,0 +1,141 @@
+	default:
+		ret = -ENODATA;
+		break;
+	}
```
same deal as `lt3074_write_word_data()`

nunojsa commented on drivers/hwmon/pmbus/lt3074.c:
```diff
@@ -0,0 +1,141 @@
+static const struct regulator_desc lt3074_reg_desc[] = {
+	PMBUS_REGULATOR("vout", 0),
+};
+#endif
```
use `__maybe_unused` instead of the preprocessor

nunojsa reviewed (commented):
Just some minor notes. Please sent this upstream

cencarna commented:
v2
- Removed extra lines
- `return` immediately for all cases in `write_word_data` and `read_word_data`
- Used `__maybe_unused`
- Used `dev_err_probe` and add extra error handling for failed read transactions (`ret < 0`)

nunojsa commented:
Note that at this point I'm expecting this to be sent upstream...

MDGrobelny commented:
Hi. When a driver will be released to main drivers tree? https://github.com/analogdevicesinc/linux/tree/main/drivers/hwmon/pmbushttps://github.com/analogdevicesinc/linux/tree/main/drivers/hwmon/pmbus 

And do we plan to release non-OS driver as well?
https://github.com/analogdevicesinc/no-OS/tree/main/drivers/power/

cencarna commented:
Upstreaming for this driver is already in progress. I think this will be released in this tree once the upstreamed patches are applied and merged to the mainline repository.

For no-OS, this will be released through here: https://github.com/analogdevicesinc/no-OS/pull/2408

nunojsa commented:
@cencarna I see this was applied upstream. Can you cherry pick the patches and update this?

cencarna commented:
v3: Series applied upstream. Added ADI-tree compliant version.

dt-bindings:
- Added separate entry for lt3074
- For this kernel version, ADI-tree example overlay which is nested and expects a "regulators" subnode is added. FYI, this series introduced an RFC to add new macros in pmbus header for "simplified" and non-nested pmbus regulators. This can be updated once that is added.

driver:
- added __maybe_unused attribute to lt3074_of_match

Note:
- Microblaze test failed. All other tests are fine. Please me know if there's anything I should to pass that CI job. Thanks.
